src/v/cloud_storage/segment_meta_cstore.h

```diff
@@ -0,0 +1,203 @@
+            return _pos == _size;
+        }
+        return other._pos == _pos && other._size == _size;
+    }
+
+    std::array<value_t, buffer_depth> _read_buf{};
+    std::array<value_t, buffer_depth> _head{};
+    std::optional<decoder_t> _decoder{std::nullopt};
+    uint32_t _pos{0};
+    uint32_t _size{0};
+};
+
+template<class value_t, class delta_t = details::delta_xor>
+class frame {
+    constexpr static uint32_t buffer_depth = details::FOR_buffer_depth;
+    constexpr static uint32_t index_mask = buffer_depth - 1;
+    using encoder_t = deltafor_encoder<value_t, delta_t>;
+    using decoder_t = deltafor_decoder<value_t, delta_t>;
+
+    struct index_value {
+        size_t ix;
+        value_t value;
+    };
+
+public:
+    using const_iterator = frame_iterator_impl<value_t, decoder_t>;
+
+    explicit frame(delta_t d)
+      : _delta_alg(std::move(d)) {}
+
+    void append(value_t value) {
+        auto ix = index_mask & _size++;
+        _head.at(ix) = value;
+        if ((_size & index_mask) == 0) {
+            if (!_tail.has_value()) {
+                _tail.emplace(_head.at(0), _delta_alg);
+            }
+            _tail->add(_head);
+        }
+    }
+
+    size_t size() const { return _size; }
+
+    const_iterator begin() const {
+        if (_tail.has_value()) {
+            decoder_t decoder(
+              _tail->get_initial_value(),
+              _tail->get_row_count(),
+              _tail->share(),
+              _delta_alg);
+            return const_iterator(std::move(decoder), _head, _size);
+        } else if (_size != 0) {
+            // special case, data is only stored in the buffer
+            // not in the compressed column
+            decoder_t decoder(0, 0, iobuf(), _delta_alg);
+            return const_iterator(std::move(decoder), _head, _size);
+        }
+        return end();
+    }
+
+    const_iterator end() const { return const_iterator(); }
+
+    const_iterator find(value_t value) const {
+        return pred_search<std::equal_to<value_t>>(value);
+    }
+
+    const_iterator upper_bound(value_t value) const {
+        return pred_search<std::greater<value_t>>(value);
+    }
+
+    const_iterator lower_bound(value_t value) const {
+        return pred_search<std::greater_equal<value_t>>(value);
+    }
+
+    std::optional<value_t> last_value() const {
+        if (_size == 0) {
+            return std::nullopt;
+        }
+        auto ix = (_size - 1) & index_mask;
+        return _head.at(ix);
+    }
+
+    bool contains(value_t value) const { return find(value) != end(); }
+
+private:
+    template<class PredT>
+    const_iterator pred_search(value_t value) const {
+        PredT pred;
+        for (auto it = begin(); it != end(); ++it) {
+            if (pred(*it, value)) {
+                return it;
+            }
+        }
+        return end();
+    }
+
+    std::array<value_t, buffer_depth> _head{};
+    mutable std::optional<encoder_t> _tail{std::nullopt};
```

Comment:
can you push `mutable` down into `encoder::share` so that it isn't at this level?

Reply:
done